### PR TITLE
Fix join plan invalidations by later transforms

### PR DESCRIPTION
**Pull request by ggevay**
Fixes https://github.com/MaterializeInc/materialize/issues/17993 by moving `JoinImplementation` to be after that fixpoint loop which has some other transforms that can invalidate a join plan.

**Unfortunately, it introduces some minor plan regressions, in that some `IS NOT NULL` predicates appear in many plans, but I don't know of any easy fix for these regressions, and the regressions are minor, so I think we should just accept them for now. For later, I opened an issue with an idea how to fix the IS NOT NULL regressions by making ColumnKnowledge smarter: https://github.com/MaterializeInc/materialize/issues/18051 (This idea would also make the IS NOT NULL removals more robust in some other situations, not just for these regressions.)**

The first commit is the main point of the PR, moving `JoinImplementation`, while the second commit fixes some plan regressions by adding a little canonicalization at the end of `JoinImplementation`, and explains the situation in the big comment at the beginning of the physical optimizer. There are full slt rewrites in both commits, so that it is possible to see what we gain by the canonicalization in the second commit.

Also, the second commit moves `LiteralConstraints`, but this change didn't change any slt plans.

### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/17993

### Tips for reviewer

- Hide whitespace, because some large plan fragments have new Filters appearing before them, which changes indentation.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: test/sqllogictest/joins.slt [hunk 1]
```diff
@@ -1023,7 +1023,7 @@ Explained Query:
     Map (4) // { arity: 2 }
       CrossJoin type=differential // { arity: 1 }
         implementation
-          %0[×]A » %1[×]A
+          %0[×] » %1[×]
```
Comment (ggevay) [inline test/sqllogictest/joins.slt:1026]: Not a regression. The new characteristics are more accurate.

**Comment by def-** [comment]
Sure, I'll trigger nightly with SQLsmith on it

**Comment by philip-stoev** [comment]
I am going to do an RQG run today.

File: src/transform/src/lib.rs [hunk 2]
```diff
@@ -471,9 +471,18 @@ impl Optimizer {
             // - there should be no RelationCSE between this LiteralConstraints and
             //   JoinImplementation, because that could move an IndexedFilter behind a Get.
             // - The last RelationCSE before JoinImplementation should be with inline_mfp = true.
-            // - Currently, JoinImplementation has to be in the same fixpoint loop with
+            // - Currently, JoinImplementation can't be before the fixpoint loop that has
```
Comment (aalexandrov) [inline src/transform/src/lib.rs:474]: Nit: can you just say 
> can't be before LiteralLifting
here instead (whether the latter is a stand-alone or a fixpoint loop seems irrelevant).

**Comment by philip-stoev** [comment]
RQG Item No 1. This query has an extra `IS NULL` check in this branch which appears redundant:

```
CREATE TABLE t2 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
CREATE INDEX t2i1 ON t2(f1);
CREATE MATERIALIZED VIEW pk2 AS SELECT DISTINCT ON (f1) f1 , f2 FROM t2 WHERE f1 IS NOT NULL AND f2 IS NOT NULL;
EXPLAIN  SELECT *  FROM t2 ,  pk2 WHERE t2.f1  = pk2.f1;
```

```
--- 16875.out	2023-03-13 11:11:04.875850133 +0100
+++ 6875.out	2023-03-13 11:11:04.835849350 +0100
@@ -1,11 +1,12 @@
 Explained Query:
   Project (#0, #1, #0, #3)
+    Filter (#0) IS NOT NULL <- THIS ONE HERE
     Join on=(#0 = #2) type=differential
       ArrangeBy keys=[[#0]]
         Get materialize.public.t2
       ArrangeBy keys=[[#0]]
         Get materialize.public.pk2
```

File: src/transform/src/lib.rs [hunk 3]
```diff
@@ -471,9 +471,18 @@ impl Optimizer {
...
             //   LiteralLifting, because the latter sometimes creates `Unimplemented` joins
             //   (despite LiteralLifting already having been run in the logical optimizer).
+            // - Not running ColumnKnowledge in the same fixpoint loop with JoinImplementation
+            //   is slightly hurting our plans. However, I'd say we should fix these problems by
+            //   making ColumnKnowledge (and/or JoinImplementation) smarter (#18051), rather than
```
Comment (aalexandrov) [inline src/transform/src/lib.rs:479]: I'm tracking this as a stretch goal for #17915.

**Comment by ggevay** [comment]
Unfortunately, there will be lots of queries with new IS NOT NULL checks from this change. [I have a plan for solving this](https://github.com/MaterializeInc/materialize/issues/18051). In the meantime, I hope these would cause only a very minor slowdown.

**Review by aalexandrov — APPROVED** [review]
In general LGTM. I don't have the full context to decide which bullet we want to bite:

1. Living with the possibility of #17993.
2. Living with potential regressions of having some `IS NOT NULL` filters not properly cleaned up.

In general I think (2) is better.

File: src/transform/src/lib.rs (hunk 2 above)
Comment (ggevay) [reply src/transform/src/lib.rs]: Fixed

**Comment by ggevay** [comment]
> I don't have the full context to decide which bullet we want to bite:
> 1. Living with the possibility of https://github.com/MaterializeInc/materialize/issues/17993.
> 2. Living with potential regressions of having some IS NOT NULL filters not properly cleaned up.

I would definitely vote for 2. It can't be more than a few percent (unless it messes up plans in some other ways, which would hopefully be revealed by @philip-stoev's RQG run), and [it's more important to be correct than efficient](https://www.notion.so/materialize/Compute-team-mission-082dab8fa9614769a3f80f024824750f).

**Comment by philip-stoev** [comment]
RQG Item No 2. Our detection of "impossible predicate" queries has always been hit-and-miss , depending on the presence of joins, the order of the predicates, the presence of parasitic `TRUE`s , the phases of the moon.

However, in this case this branch fails to detect this totally realistic and plausible query as a constant query, and it was detected as such in the parent commit. So I think it should be treated as a regression.

```

DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE lineitem (
    l_orderkey       integer 
);

CREATE TABLE orders (
    o_orderkey       integer NOT NULL
);


EXPLAIN SELECT * FROM orders LEFT JOIN lineitem ON ( TRUE ) WHERE o_orderkey = 1349 AND o_orderkey = 173 ;
```

**Comment by philip-stoev** [comment]
Item No 3B . I think this is the same as before, except that the parent commit also does an incomplete job of convering the query to Constant:

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE lineitem (
    l_orderkey       integer 
);

CREATE TABLE orders (
    o_orderkey       integer NOT NULL,
    o_custkey        integer NOT NULL
);

CREATE MATERIALIZED VIEW orders_pk AS SELECT DISTINCT ON (o_orderkey) o_orderkey, o_custkey FROM orders;

EXPLAIN SELECT * FROM orders_pk LEFT JOIN lineitem
ON ( l_orderkey = o_orderkey )
WHERE o_custkey = 33 AND o_custkey = 333;
```

**Comment by philip-stoev** [comment]
> Unfortunately, it introduces some minor plan regressions, in that some IS NOT NULL

I should start reading the descriptions of the PRs that I am testing before I report issues against them.
